### PR TITLE
ARROW-6649: [R] print methods for Array, ChunkedArray, Table, RecordBatch

### PR DESCRIPTION
**Pull request by nealrichardson**
See the readme and the new tests for example output. 

This patch also fixes a validation bug in `dictionary()`, aliases that to `DictionaryType$create`, and adds default arguments.

### CODE REVIEW

File: r/tests/testthat/test-RecordBatch.R [hunk 1]
```diff
@@ -120,6 +120,23 @@ test_that("head and tail on RecordBatch", {
...
+  expect_output(
+    print(batch),
+    paste(
+      "RecordBatch",
+      "10 rows x 5 columns",
+      "$int <int32>",
+      "$dbl <double>",
+      "$lgl <bool>",
+      "$chr <string>",
+      "$fct <dictionary<values=string, indices=int8, ordered=0>>",
```
Comment (pitrou) [inline r/tests/testthat/test-RecordBatch.R]: Just for the record, R doesn't have boolean values? ("ordered=0")

File: r/R/dictionary.R [hunk 2]
```diff
@@ -39,20 +39,21 @@ DictionaryType <- R6Class("DictionaryType",
...
+  assert_is(value_type, "DataType")
+  shared_ptr(DictionaryType, DictionaryType__initialize(index_type, value_type, ordered))
+}
 
 #' Create a dictionary type
 #'
-#' @param index_type index type, e.g. [int32()]
-#' @param value_type value type, probably [utf8()]
-#' @param ordered Is this an ordered dictionary ?
+#' @param index_type A DataType for the indexes (default [int32()])
```
Comment (bkietz) [inline r/R/dictionary.R]: nit:
```suggestion
#' @param index_type A DataType for the indices (default [int32()])
```

**Review by bkietz — APPROVED** [review]

File: r/tests/testthat/test-RecordBatch.R (hunk 1 above)
Comment (nealrichardson) [reply r/tests/testthat/test-RecordBatch.R]: It does, but this is coming straight from the C++ `ToString` method. I guess I could munge it in R.

Comment (pitrou) [reply r/tests/testthat/test-RecordBatch.R]: I suppose it's fine then.

File: r/man/dictionary.Rd [hunk 3]
```diff
@@ -4,14 +4,15 @@
 \alias{dictionary}
 \title{Create a dictionary type}
 \usage{
-dictionary(index_type, value_type, ordered = FALSE)
+dictionary(index_type = int32(), value_type = utf8(),
+  ordered = FALSE)
 }
 \arguments{
-\item{index_type}{index type, e.g. \code{\link[=int32]{int32()}}}
+\item{index_type}{A DataType for the indexes (default \code{\link[=int32]{int32()}})}
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/man/dictionary.Rd]:
```suggestion
\item{index_type}{A DataType for the indices (default \code{\link[=int32]{int32()}})}
```

File: r/R/dictionary.R [hunk 4]
```diff
@@ -39,20 +39,21 @@ DictionaryType <- R6Class("DictionaryType",
     ordered = function() DictionaryType__ordered(self)
   )
 )
+DictionaryType$create <- function(index_type = int32(),
+                                  value_type = utf8(),
```
[the remark this answers is not in the document]
Comment (nealrichardson) [reply r/R/dictionary.R]: Maybe so, but this is what an R `factor` type gets translated to, so it seemed reasonable.

File: r/tests/testthat/test-RecordBatch.R (hunk 1 above)
Comment (nealrichardson) [reply r/tests/testthat/test-RecordBatch.R]: I added some munging in https://github.com/apache/arrow/pull/5492/commits/02afb89f24dccecd88176d483bb406c82908c781#diff-63df045aa8aff63fdd460b816987ba0dR65-R69

**Comment by nealrichardson** [comment]
Travis passed here: https://travis-ci.org/nealrichardson/arrow

**Pull request closed without merging**